### PR TITLE
[Feature] Support MTP speculative decoding for Bailing hybrid models

### PR DESCRIPTION
**Pull request by alex101-ops**
## Purpose

Add MTP (Multi-Token Prediction) speculative decoding support for Bailing MoE hybrid models, and enhance the linear attention backend to support speculative decode with CUDAGraph.

**Key changes:**

- **New `BailingMoeV25MTPModel`**: Add the draft model implementation for Bailing MoE V2.5 MTP speculative decoding (`bailing_moe_mtp.py`), including `BailingMoeV25MultiTokenPredictor` with shared embedding/lm_head support and MoE expert weight loading.
- **Speculative config**: Register `bailing_hybrid_mtp` model type and auto-detect Bailing architectures to override `model_type` and `architectures` for MTP draft model construction.
- **Model registry**: Register `BailingMoeV25MTPModel` in `_SPECULATIVE_DECODING_MODELS`.
- **Model arch config convertor**: Add `BailingHybridMTPModelArchConfigConvertor` to provide correct `num_hidden_layers` for MTP layers.
- **Linear attention backend**: Enhance `LinearAttentionMetadataBuilder` to support speculative decode by splitting decode/prefill state indices, handling multi-token draft indices (`num_accepted_tokens`, `query_start_loc_d`), and CUDAGraph capture with spec decode awareness. Report `UNIFORM_BATCH` CUDAGraph support for Bailing hybrid models.
- **MiniMax & Bailing linear attention kernels**: Add `_linear_attention_decode_spec` triton kernel for speculative decode, and update `linear_attention_prefill_and_mix` to use split decode/prefill state indices.
- **MTP proposer**: Support `word_embeddings` attribute in target model embedding lookup, gracefully handle `lm_head` location across model nesting levels, and make `topk_indices_buffer` sharing robust to different model structures.
- **Eagle utils**: Support `word_embeddings` as a fallback embedding attribute and look for `lm_head` on the inner language model when not found on the outer model.
- **GPU model runner**: Include `LinearAttentionMetadataBuilder` in the spec decode metadata path.

## Test Plan

Run the following unit tests:

```bash
# Bailing MTP config tests
pytest tests/config/test_bailing_mtp_config.py -v

# MTP proposer tests (includes Bailing word_embeddings / shared_head sharing test)
pytest tests/v1/spec_decode/test_mtp.py -v

# Linear attention metadata builder tests (CUDAGraph support, spec decode metadata)
pytest tests/v1/attention/test_linear_attention_metadata_builder.py -v
```

## Test Result

All newly added unit tests pass:
- `test_bailing_hybrid_mtp_hf_config_override` — verifies `bailing_hybrid` → `bailing_hybrid_mtp` override
- `test_bailing_hybrid_mtp_model_arch_config` — verifies MTP layer count and architecture config
- `test_bailing_mtp_load_model_shares_word_embeddings_and_shared_heads` — verifies embedding/lm_head/buffer sharing between target and draft model
- `test_bailing_linear_attention_reports_uniform_batch_cudagraph_support` — verifies CUDAGraph support level for Bailing
- `test_non_bailing_linear_attention_keeps_single_token_cudagraph_support` — verifies non-Bailing models are unaffected

hardware: H800
MODEL: Ling 2.6-flash
image

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by alex101-ops** [comment]
@ZJY0516  I have completed the end-to-end verification of this feature on NVIDIA GPUs such as H800. Could you help review it?

File: vllm/v1/attention/backends/linear_attn.py [hunk 1]
```diff
@@ -58,28 +92,173 @@ def __init__(
...
+        )
+        self._init_reorder_batch_threshold(1, self.use_spec_decode)
+
+    def build_for_cudagraph_capture(
+        self, common_attn_metadata: CommonAttentionMetadata
+    ) -> LinearAttentionMetadata:
+        num_accepted_tokens = None
+        if self.use_spec_decode:
+            assert common_attn_metadata.max_query_len <= 1 + self.num_spec_tokens, (
+                "Linear attention only supports decode-only full CUDAGraph "
```
Comment (ZJY0516) [inline vllm/v1/attention/backends/linear_attn.py]: This log is a little confusing

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -2355,9 +2356,14 @@ def _build_attn_group_metadata(
...
-                builder, (Mamba2AttentionMetadataBuilder, GDNAttentionMetadataBuilder)
+                builder,
+                (
+                    Mamba2AttentionMetadataBuilder,
+                    GDNAttentionMetadataBuilder,
+                    LinearAttentionMetadataBuilder,
+                ),
             ):
-                assert ubid is None, "UBatching not supported with GDN yet"
+                assert ubid is None, "UBatching not supported with spec decode metadata yet"
```
Comment (ZJY0516) [inline vllm/v1/worker/gpu_model_runner.py]:
```suggestion
                assert ubid is None, "UBatching not supported with hybrid model spec decode metadata yet"
```

File: vllm/v1/spec_decode/llm_base_proposer.py [hunk 3]
```diff
@@ -1428,17 +1431,25 @@ def _maybe_share_lm_head(self, target_language_model: nn.Module) -> None:
...
+        draft_inner_model = getattr(self.model, "model", None)
+        if (
+            target_inner_model is not None
+            and draft_inner_model is not None
+            and hasattr(target_inner_model, "topk_indices_buffer")
+        ):
+            target_buffer = target_inner_model.topk_indices_buffer
+            if hasattr(draft_inner_model, "topk_indices_buffer"):
+                del draft_inner_model.topk_indices_buffer
+            draft_inner_model.topk_indices_buffer = target_buffer
```
Comment (ZJY0516) [inline vllm/v1/spec_decode/llm_base_proposer.py]: Could you explain why we need to change this?

**Comment by ZJY0516** [comment]
Could you also provide acceptance rate during test?

File: vllm/v1/spec_decode/llm_base_proposer.py (hunk 3 above)
Comment (alex101-ops) [reply vllm/v1/spec_decode/llm_base_proposer.py]: Avoid throwing AttributeError directly when some model do not have the. model attribute.

**Comment by alex101-ops** [comment]
> Could you also provide acceptance rate during test?
ok，here are the benchmark test results
1780983608922_192f8c13e89a4a88ad9fcb8801587609

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (alex101-ops) [reply vllm/v1/worker/gpu_model_runner.py]: okok, i replaced it

**Comment by ZJY0516** [comment]
@claude review

File: vllm/v1/spec_decode/llm_base_proposer.py (hunk 3 above)
Comment (ZJY0516) [reply vllm/v1/spec_decode/llm_base_proposer.py]: Which model?

File: vllm/v1/attention/backends/linear_attn.py [hunk 4]
```diff
@@ -58,28 +91,174 @@ def __init__(
...
+                torch.where(
+                    is_padded_decode,
+                    torch.full_like(
+                        legacy_state_indices_tensor[:num_decodes],
+                        PAD_SLOT_ID,
+                    ),
+                    legacy_state_indices_tensor[:num_decodes],
+                ),
+                non_blocking=True,
+            )
```
Comment (ZJY0516) [inline vllm/v1/attention/backends/linear_attn.py]: Why we have 2 `self.decode_legacy_state_indices_tensor[:num_decodes].copy_` here?

File: vllm/v1/attention/backends/linear_attn.py [hunk 5]
```diff
@@ -58,28 +91,174 @@ def __init__(
...
     def build(
         self,
         common_prefix_len: int,
         common_attn_metadata: CommonAttentionMetadata,
         fast_build: bool = False,
+        *,
+        num_accepted_tokens: torch.Tensor | None = None,
+        num_decode_draft_tokens_cpu: torch.Tensor | None = None,
     ) -> LinearAttentionMetadata:
+        del common_prefix_len, fast_build, num_decode_draft_tokens_cpu
```
Comment (ZJY0516) [inline vllm/v1/attention/backends/linear_attn.py]: Why delete this?

File: vllm/model_executor/models/bailing_moe_mtp.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+    def forward(
+        self,
+        input_ids: torch.Tensor | None,
+        positions: torch.Tensor,
+        hidden_states: torch.Tensor,
+        intermediate_tensors: IntermediateTensors | None = None,
+        inputs_embeds: torch.Tensor | None = None,
+        spec_step_idx: int = 0,
+    ) -> torch.Tensor:
+        del intermediate_tensors
```
Comment (ZJY0516) [inline vllm/model_executor/models/bailing_moe_mtp.py]: same question

File: vllm/v1/worker/gpu/spec_decode/eagle/utils.py [hunk 7]
```diff
@@ -45,8 +45,10 @@ def load_eagle_model(target_model: nn.Module, vllm_config: VllmConfig) -> nn.Mod
 
     # Skip embedding sharing under PP — each rank owns its own embedding.
     if get_pp_group().world_size == 1:
-        target_embed = getattr(target_inner, "embed_tokens", None) or getattr(
-            target_inner, "embedding", None
+        target_embed = (
+            getattr(target_inner, "embed_tokens", None)
+            or getattr(target_inner, "embedding", None)
+            or getattr(target_inner, "word_embeddings", None)
```
Comment (zixi-qi) [inline vllm/v1/worker/gpu/spec_decode/eagle/utils.py]: Same as above

File: vllm/v1/worker/gpu/spec_decode/eagle/utils.py [hunk 8]
```diff
@@ -57,6 +59,8 @@ def load_eagle_model(target_model: nn.Module, vllm_config: VllmConfig) -> nn.Mod
             draft_inner.embed_tokens = target_embed
 
     target_lm_head = getattr(target_model, "lm_head", None)
+    if target_lm_head is None:
+        target_lm_head = getattr(target_language_model, "lm_head", None)
```
Comment (zixi-qi) [inline vllm/v1/worker/gpu/spec_decode/eagle/utils.py]: Could this plumbing happen on the model definition side?

File: vllm/v1/spec_decode/llm_base_proposer.py [hunk 9]
```diff
@@ -1305,9 +1305,12 @@ def _maybe_share_embeddings(self, target_language_model: nn.Module) -> None:
                 target_embed_tokens = inner_model.embed_tokens
             elif hasattr(inner_model, "embedding"):
                 target_embed_tokens = inner_model.embedding
+            elif hasattr(inner_model, "word_embeddings"):
+                target_embed_tokens = inner_model.word_embeddings
```
Comment (zixi-qi) [inline vllm/v1/spec_decode/llm_base_proposer.py]: Would it be possible to use something like "embed_tokens" instead of "word_embeddings" so that the shared path don't have to be changed?

Comment (alex101-ops) [reply vllm/v1/spec_decode/llm_base_proposer.py]: Ok. I removed the `word_embeddings` branch from the shared proposer path and added an `embed_tokens` property to the Bailing model instead, so the existing embedding sharing logic can be reused unchanged.

File: vllm/v1/worker/gpu/spec_decode/eagle/utils.py (hunk 8 above)
Comment (alex101-ops) [reply vllm/v1/worker/gpu/spec_decode/eagle/utils.py]: OK,Fixed.

File: vllm/v1/worker/gpu/spec_decode/eagle/utils.py (hunk 7 above)
Comment (alex101-ops) [reply vllm/v1/worker/gpu/spec_decode/eagle/utils.py]: OK,Fixed.

File: vllm/model_executor/models/bailing_moe_mtp.py (hunk 6 above)
Comment (alex101-ops) [reply vllm/model_executor/models/bailing_moe_mtp.py]: fixed it

File: vllm/v1/attention/backends/linear_attn.py (hunk 4 above)
Comment (alex101-ops) [reply vllm/v1/attention/backends/linear_attn.py]: The first copy is redundant for the legacy path, so I removed it in the Bailing-specific builder. The 2-D spec-decode path still uses its separate buffer.

**Comment by alex101-ops** [comment]
Thanks for the feedback. I have updated the PR to address the review comments. The redundant code has been removed, and the Bailing-specific changes were moved out of the shared/common paths where possible, including the embedding sharing path and the linear attention helper logic.

I also reran performance and accuracy tests on H200 after these changes, and did not observe any issues.

Could you please take another look when you have time and let me know if this is ready to merge?
@ZJY0516 @zixi-qi 

gsm8k
1782726332079_b7bc809b57184a20a382bd7aab79aa33
benchmark test results
1782726615388_8550c09eabf1442bb53903762c6ad932

**Review by ZJY0516 — APPROVED** [review]

**Comment by zixi-qi** [comment]
LGTM

**Comment by alex101-ops** [comment]
@ZJY0516 @zixi-qi All CI checks have passed now. Could you please help merge this PR?
